File: frontend/src/api/platform.ts
/**
 * Typed client for LunaMatch platform APIs (/api/v1).
 * Uses same-origin `/api` (Vite proxy / nginx / StripApiPrefixMiddleware).
 */

const API_BASE = (import.meta.env.VITE_API_BASE as string | undefined)?.replace(/\/$/, "") || "/api";

async function request(path: string, init?: RequestInit): Promise<Response> {
  const url = `${API_BASE}${path.startsWith("/") ? path : `/${path}`}`;
  try {
    return await fetch(url, init);
  } catch {
    throw new Error("Failed to reach LunaMatch platform API");
  }
}

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const text = await res.text();
    try {
      const data = JSON.parse(text) as { detail?: unknown };
      if (data?.detail != null) {
        throw new Error(typeof data.detail === "string" ? data.detail : JSON.stringify(data.detail));
      }
    } catch (err) {
      if (err instanceof Error && err.message && err.message !== text) throw err;
    }
    throw new Error(text || `Request failed (${res.status})`);
  }
  return res.json() as Promise<T>;
}

export type Project = {
  id: string;
  name: string;
  description: string;
  meta: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type Dataset = {
  id: string;
  project_id: string;
  name: string;
  kind: string;
  content_type: string;
  byte_size: number;
  object_key: string;
  data_status: string;
  meta: Record<string, unknown>;
  created_at: string;
  download_url?: string | null;
};

export type AnalysisRun = {
  id: string;
  project_id: string;
  kind: string;
  status: string;
  progress: number;
  message: string;
  engine: string;
  params: Record<string, unknown>;
  lineage: Record<string, unknown>;
  error?: string | null;
  created_at: string;
  updated_at: string;
};

export type RegistrationResult = {
  id: string;
  run_id: string;
  engine: string;
  engine_version: string;
  device: string;
  fallback_used: boolean;
  runtime_ms?: number | null;
  metrics: Record<string, unknown>;
  reliability: Record<string, unknown>;
  object_keys: Record<string, string>;
  limitations: string;
  download_urls: Record<string, string>;
};

export type ExportJob = {
  id: string;
  run_id: string;
  status: string;
  object_key?: string | null;
  byte_size: number;
  download_url?: string | null;
  error?: string | null;
  created_at: string;
};

export type PlatformHealth = {
  status: string;
  service: string;
  database: string;
  redis: string;
  minio: string;
  matchers: Array<{ engine: string; available: boolean; status: string }>;
};

export async function platformHealth(): Promise<PlatformHealth | null> {
  try {
    const res = await request("/v1/health");
    if (!res.ok) return null;
    return (await res.json()) as PlatformHealth;
  } catch {
    return null;
  }
}

export async function listProjects() {
  return json<Project[]>(await request("/v1/projects"));
}

export async function createProject(name: string, description = "") {
  return json<Project>(
    await request("/v1/projects", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ name, description }),
    }),
  );
}

export async function listDatasets(projectId: string) {
  return json<Dataset[]>(await request(`/v1/datasets/project/${projectId}`));
}

export async function uploadDatasetFile(
  projectId: string,
  file: File,
  opts?: { kind?: string; dataStatus?: string },
) {
  const init = await json<{
    dataset_id: string;
    upload_url: string;
    headers: Record<string, string>;
  }>(
    await request("/v1/datasets/uploads/init", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        project_id: projectId,
        filename: file.name,
        content_type: file.type || "application/octet-stream",
        byte_size: file.size,
        kind: opts?.kind ?? "optical",
        data_status: opts?.dataStatus ?? "real uploaded image",
      }),
    }),
  );

  const put = await fetch(init.upload_url, {
    method: "PUT",
    headers: { "Content-Type": file.type || "application/octet-stream", ...(init.headers || {}) },
    body: file,
  });
  if (!put.ok) throw new Error(`MinIO upload failed (${put.status})`);

  return json<Dataset>(
    await request("/v1/datasets/uploads/confirm", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ dataset_id: init.dataset_id }),
    }),
  );
}

export async function startRegistrationRun(body: {
  project_id: string;
  reference_dataset_id: string;
  source_dataset_id: string;
  engine?: string;
  allow_fallback?: boolean;
}) {
  return json<AnalysisRun>(
    await request("/v1/registration/runs", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        engine: "akaze",
        allow_fallback: true,
        ...body,
      }),
    }),
  );
}

export async function getRegistrationRun(runId: string) {
  return json<AnalysisRun>(await request(`/v1/registration/runs/${runId}`));
}

export async function getRegistrationResults(runId: string) {
  return json<RegistrationResult>(await request(`/v1/registration/runs/${runId}/results`));
}

export async function cancelRegistrationRun(runId: string) {
  return json<AnalysisRun>(await request(`/v1/registration/runs/${runId}/cancel`, { method: "POST" }));
}

export async function listRegistrationRuns(projectId: string) {
  return json<AnalysisRun[]>(await request(`/v1/registration/projects/${projectId}/runs`));
}

export async function createExport(runId: string) {
  return json<ExportJob>(
    await request("/v1/exports", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ run_id: runId }),
    }),
  );
}

export async function getExport(exportId: string) {
  return json<ExportJob>(await request(`/v1/exports/${exportId}`));
}

export async function persistIceAnalysis(body: Record<string, unknown>) {
  return json<AnalysisRun>(
    await request("/v1/ice/runs", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }),
  );
}

export function pollUntil<T>(
  fn: () => Promise<T>,
  done: (value: T) => boolean,
  opts?: { intervalMs?: number; timeoutMs?: number },
): Promise<T> {
  const intervalMs = opts?.intervalMs ?? 1500;
  const timeoutMs = opts?.timeoutMs ?? 180_000;
  const start = Date.now();
  return new Promise((resolve, reject) => {
    const tick = async () => {
      try {
        const value = await fn();
        if (done(value)) {
          resolve(value);
          return;
        }
        if (Date.now() - start > timeoutMs) {
          reject(new Error("Timed out waiting for job"));
          return;
        }
        window.setTimeout(() => void tick(), intervalMs);
      } catch (err) {
        reject(err);
      }
    };
    void tick();
  });
}

File: backend/app/schemas/platform.py
"""Pydantic schemas for /api/v1."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


class ORMModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class ProjectCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str = ""
    meta: dict[str, Any] = Field(default_factory=dict)


class ProjectUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = None
    meta: Optional[dict[str, Any]] = None


class ProjectOut(ORMModel):
    id: uuid.UUID
    name: str
    description: str
    meta: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class UploadInitRequest(BaseModel):
    project_id: uuid.UUID
    filename: str
    content_type: str
    byte_size: int
    kind: str = "optical"
    data_status: str = "real uploaded image"
    meta: dict[str, Any] = Field(default_factory=dict)


class UploadInitResponse(BaseModel):
    dataset_id: uuid.UUID
    object_key: str
    upload_url: str
    expires_in: int
    headers: dict[str, str] = Field(default_factory=dict)


class UploadConfirmRequest(BaseModel):
    dataset_id: uuid.UUID
    checksum_sha256: Optional[str] = None


class DatasetOut(ORMModel):
    id: uuid.UUID
    project_id: uuid.UUID
    name: str
    kind: str
    content_type: str
    byte_size: int
    object_key: str
    data_status: str
    meta: dict[str, Any]
    created_at: datetime
    download_url: Optional[str] = None


class RegistrationRunCreate(BaseModel):
    project_id: uuid.UUID
    reference_dataset_id: uuid.UUID
    source_dataset_id: uuid.UUID
    engine: str = "akaze"
    allow_fallback: bool = True
    preprocessing: dict[str, Any] = Field(default_factory=lambda: {"clahe": True})
    ransac: dict[str, Any] = Field(default_factory=lambda: {"reproj_threshold_px": 3.0})
    meta: dict[str, Any] = Field(default_factory=dict)


class AnalysisRunOut(ORMModel):
    id: uuid.UUID
    project_id: uuid.UUID
    kind: str
    status: str
    progress: float
    message: str
    engine: str
    params: dict[str, Any]
    lineage: dict[str, Any]
    error: Optional[str]
    created_at: datetime
    updated_at: datetime
    started_at: Optional[datetime]
    finished_at: Optional[datetime]


class RegistrationResultOut(ORMModel):
    id: uuid.UUID
    run_id: uuid.UUID
    engine: str
    engine_version: str
    device: str
    fallback_used: bool
    runtime_ms: Optional[float]
    metrics: dict[str, Any]
    reliability: dict[str, Any]
    object_keys: dict[str, Any]
    limitations: str
    download_urls: dict[str, str] = Field(default_factory=dict)


class IcePersistRequest(BaseModel):
    project_id: uuid.UUID
    thresholds: dict[str, Any] = Field(default_factory=dict)
    cluster_metrics: dict[str, Any] = Field(default_factory=dict)
    volume_scenarios: dict[str, Any] = Field(default_factory=dict)
    landing_candidates: list[dict[str, Any]] = Field(default_factory=list)
    rover_routes: list[dict[str, Any]] = Field(default_factory=list)
    object_keys: dict[str, Any] = Field(default_factory=dict)
    meta: dict[str, Any] = Field(default_factory=dict)


class ExportCreate(BaseModel):
    run_id: uuid.UUID
    include: list[str] = Field(default_factory=lambda: ["imagery", "metrics", "geojson", "readme"])


class ExportOut(ORMModel):
    id: uuid.UUID
    run_id: uuid.UUID
    status: str
    object_key: Optional[str]
    byte_size: int
    meta: dict[str, Any]
    error: Optional[str]
    download_url: Optional[str] = None
    created_at: datetime


class HealthOut(BaseModel):
    status: str
    service: str
    app_env: str
    database: str
    redis: str
    minio: str
    matchers: list[dict[str, Any]]
